controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+                        "message" => ?message,
+                        "peer" => peer,
+                    );
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+async fn send_protocol_error(
+    log: &Logger,
+    socket: &UdpSocket,
+    peer: &SocketAddr,
+    header: &Header,
+    modules: &ModuleId,
+) {
+    error!(log, "Protocol error");
+    let err = MessageError::ProtocolError;
+    let message = Message {
+        header: *header,
+        modules: *modules,
+        body: MessageBody::HostResponse(HostResponse::Error(err)),
+    };
+    let mut buf = [0u8; Message::MAX_SIZE];
+    hubpack::serialize(&mut buf, &message).unwrap();
+    match socket.send_to(&buf, peer).await {
+        Err(e) => {
+            error!(
+                log,
+                "failed to send protocol error";
+                "reason" => ?e,
+                "peer" => peer
+            );
+        }
+        Ok(n_bytes) => assert_eq!(n_bytes, Message::MAX_SIZE),
+    }
+}
+
+async fn send_version_mismatch(
+    log: &Logger,
+    socket: &UdpSocket,
+    peer: &SocketAddr,
+    header: &Header,
+    modules: &ModuleId,
+) -> Result<(), Error> {
+    error!(
+        log,
+        "Mismatched protocol versions";
+        "expected" => message::version::V1,
+        "actual" => header.version,
+    );
+    let err = MessageError::VersionMismatch {
+        expected: message::version::V1,
+        actual: header.version,
+    };
+    let message = Message {
+        header: *header,
+        modules: *modules,
+        body: MessageBody::HostResponse(HostResponse::Error(err)),
+    };
+    let mut buf = [0u8; Message::MAX_SIZE];
+    hubpack::serialize(&mut buf, &message).unwrap();
+    let n_bytes = socket.send_to(&buf, peer).await?;
+    assert_eq!(n_bytes, Message::MAX_SIZE);
+    Ok(())
+}
+
+// Send a request from the host to the SP, returning the number of bytes sent.
+async fn send_outgoing_request(
+    socket: &UdpSocket,
+    peer: &SocketAddrV6,
+    request: &mut OutstandingHostRequest,
+    buf: &mut [u8],
+) -> Result<usize, Error> {
+    let mut msg_size = match hubpack::serialize(buf, &request.request.message) {
+        Ok(n) => n,
+        Err(e) => return Err(Error::SerDes(e)),
+    };
+    if let Some(data) = &request.request.data {
+        buf[msg_size..data.len()].copy_from_slice(data);
+        msg_size += data.len();
+    }
+    let n_bytes = socket.send_to(&buf[..msg_size], peer).await?;
+    assert_eq!(n_bytes, msg_size);
+    request.last_sent = Instant::now();
+    Ok(n_bytes)
+}
+
+async fn request_loop<H, F>(
+    log: Logger,
+    mut incoming_request_rx: mpsc::Receiver<SpRpcRequest>,
+    outgoing_response_tx: mpsc::Sender<HostRpcResponse>,
+    request_handler: H,
+) where
+    H: Fn(SpRpcRequest) -> F + Send + Sync + 'static,
+    F: Future<Output = Result<HostRpcResponse, Error>> + Send,
+{
+    loop {
+        let incoming_request = incoming_request_rx.recv().await.unwrap();
```

Comment:
Is it possible for the sender to close the channel here?

If so, we could get a clean exit for free with something like
```rust
while let Some(incoming_request) = incoming_request_rx.recv().await {
    // ...
```

Reply:
Oh, nice idea. I don't know yet if we'll ever see `None` here, since it depends on how I end up using this type in Dendrite. I.e., if we drop the sender before we abort this task. But I think your suggestion will be cleaner regardless, thanks!